### PR TITLE
When updating a BackendService copy the backends from the existing re…

### PR DESCRIPTION
**Pull request by mmamczur**
…source to avoid detachment.

### CODE REVIEW

File: pkg/backends/backends.go [hunk 1]
```diff
@@ -353,6 +353,8 @@ func (b *Backends) EnsureL4BackendService(name, hcLink, protocol, sessionAffinit
 	klog.V(2).Infof("EnsureL4BackendService: updating backend service %v", name)
 	// Set fingerprint for optimistic locking
 	expectedBS.Fingerprint = bs.Fingerprint
+	// Copy backends to avoid detaching them during update. This could be replaced with a patch call in the future.
+	expectedBS.Backends = bs.Backends
```
Comment (aojea) [inline pkg/backends/backends.go:357]: why are these things not added on line 312 when we build the desired state?

adding it here gets the risk of introducing a new change in the middle that carries the same bug

**Comment by aojea** [comment]
/lgtm
to the best of my knowledge this looks correct

/assign @bowei

File: pkg/backends/backends.go (hunk 1 above)
Comment (bowei) [inline pkg/backends/backends.go:357]: Why do we do this after line 345 where we check if it's equal? Shouldn't we do it before the equality check?

File: pkg/backends/backends_test.go [hunk 2]
```diff
@@ -111,6 +112,66 @@ func TestEnsureL4BackendService(t *testing.T) {
 	}
 }
 
+func TestEnsureL4BackendServiceDoesNotDetachBackends(t *testing.T) {
+	serviceName := "test-service"
```
Comment (bowei) [inline pkg/backends/backends_test.go:116]: use const for constants

File: pkg/backends/backends_test.go [hunk 3]
```diff
@@ -111,6 +112,66 @@ func TestEnsureL4BackendService(t *testing.T) {
...
+	}
+	if diff := cmp.Diff(backend, existingBS.Backends[0]); diff != "" {
+		t.Errorf("BackendService.Backends were changed, expected no change: %s", diff)
+	}
+	description, err := utils.MakeL4LBServiceDescription(namespacedName.String(), "", meta.VersionGA, false, utils.ILB)
+	if err != nil {
+		t.Errorf("utils.MakeL4LBServiceDescription() failed %v", err)
+	}
+	if bs.Description != description {
+		t.Errorf("BackendService.Description was not populated correctly, want=%q, got=%q", description, bs.Description)
```
Comment (bowei) [inline pkg/backends/backends_test.go:171]: got before want

**Comment by mmamczur** [comment]
check #2301 for a follow up to the comments from here

**Pull request merged**